new pinned toolbar & history view keys long press functions

codokie opened the pull request:
Fixes #686
I added new functions on long press (credit to @Roccobot for the idea):
| Key   | Tap                   | Tap + Hold                        |
|-------|-----------------------|----------------------------------|
| ᐸ     | Move cursor left      | Move cursor to the beginning of the line |
| ᐳ     | Move cursor right     | Move cursor to the end of the line       |
| ⋀     | Move cursor up        | Move cursor page up           |
| ⋁     | Move cursor down      | Move cursor page down            |
| ⎘    | Copy selection        | Copy all                             |
| ↩    | Undo                  | Redo                                 |
| ↪    | Redo                  | Undo                                 |
| ⬚    | Select word           | Select all                           |

The history view now has Up, Down, Undo, and One Handed Mode (fixes #513) keys to fill the blank space, and the long press functions are available there as well.
The Select All key has been removed to preserve space and because it is now possible to select all by long pressing the select word (you can also long press the copy key to select all + copy).
The "clear clipboard" key has been moved to avoid pressing it by accident when attempting to press the "close history" key, and in order to have the Cut key closer to the Select button.

Here are old history view keys:
![old](https://github.com/Helium314/HeliBoard/assets/151087174/7038f3ae-273a-4593-9923-17ef3e7b0af3)

and here is the new key layout:
![new](https://github.com/Helium314/HeliBoard/assets/151087174/0884eda6-562d-4511-96ca-6c06e95629d6)

I know there are plans to make the history view customizable in the future, but in the meantime I believe this should be sufficient for most people.

On a side note, the new copyAllText() that is triggered when long pressing the Copy key may copy up to Integer.MAX_VALUE * 4  bytes to memory which is a lot.
Is there a value that is more careful and still makes sense to achieve Copy All functionality ?

Note that the old copyText() used for Copy key (short press) may copy up to Integer.MAX_VALUE * 2 bytes to memory, which can also lead to memory issues if for some reason the current input field has a lot of text

Helium314 commented:
This is again mixing a bunch of small and independent changes, making review harder... I updated the issue templates to hopefully make more clear that I dislike it.

Other than that, I guess it's mostly fine.
Some issues with your code (I haven't looked through everything:
* The code for determining the keycode on long pressing a toolbar key is essentially duplicated. The keycode should be determined in a similar way as the normal code.
* You add keycodes above the `heliboard only codes` section in `KeyCode`, with number ranges that are also used in FlorisBoard. Since these codes may get added in FlorisBoard too, but likely with different actions this is prone to breaking keyboard layout compatibility.

codokie commented:
Thanks for the feedback.

I pushed a new commit [7a4e591](https://github.com/Helium314/HeliBoard/pull/691/commits/7a4e59164efcf9043c98e1a3ea664451a206673b)
I agree the code seemed repetitive, so I've moved that code to a new function for better maintainability
I did not know there is a possibility of a conflict with the keycodes. Now the new codes are placed under the app specific section.

I can revert the change to the clipboard view but it is not a major one and it partly depends on the changes of the long click functions.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/internal/keyboard_parser/floris/KeyCode.kt:
```diff
@@ -132,8 +130,10 @@ object KeyCode {
     const val ACTION_NEXT =               -10006
     const val ACTION_PREVIOUS =           -10007
+    const val MOVE_PAGE_UP =              -10008
+    const val MOVE_PAGE_DOWN =            -10009
     // Code value representing the code is not specified.
-    const val NOT_SPECIFIED =             -10008 // todo: not sure if there is need to have the "old" unspecified keyCode different, just test it and maybe merge
+    const val NOT_SPECIFIED =             -10010 // todo: not sure if there is need to have the "old" unspecified keyCode different, just test it and maybe merge
```
Actually there is no need to have the `NOT_SPECIFIED` code last, you can also just leave it at `-10008`.
Though it would only be relevant if someone would use this code in a layout, which is quite unlikely.

codokie replied:
ok, I changed it back to the old code.
I also moved the newly added `CLIPBOARD_COPY_ALL` which I somehow missed to the app specific section

Helium314 commented:
> I agree the code seemed repetitive, so I've moved that code to a new function for better maintainability

Thank, though I think there might have been a little misunderstanding.
When I wrote "_The keycode should be determined in a similar way as the normal code._" I actually meant to have something like `ToolbarUtils.getLongpressCodeForToolbarKey` analogous to `ToolbarUtils.getCodeForToolbarKey` (the function name admittedly sounds awkward and could be better).
Sorry I wasn't clear enough about this.

> I can revert the change to the clipboard view but it is not a major one and it partly depends on the changes of the long click functions.

No, I think it should be fine. Especially because it's all in a separate commit.

Helium314 commented on app/src/main/res/layout/strip_container.xml:
```diff
@@ -24,16 +24,23 @@
-            android:layout_height="0dp"
-            android:layout_weight="1" />
-    </LinearLayout>
+        android:visibility="gone"
```
Why did you add `android:visibility="gone"`? Is it necessary now for some reason?

codokie replied:
no I don't think it is necessary. I added it while debugging and probably forgot to remove it

codokie commented:
Good idea, I moved the function to the ToolbarUtils class
I also removed the onLongClickClipboardKey from SuggestionStripView because it does not really belong there and can be handled with onCodeInput()
Thanks for the feedback

Helium314 commented:
Code looks good, I hope tomorrow I finally have time for an actual test.

Helium314 commented:
The PR mostly does what it intends to. I find undo / redo on the same key rather weird, but i can imagine it's really convenient :+1: 

I found two issues though:
Pasting by long pressing the clipboard key replaces the underlined word, which is not the case without this PR.
Long pressing the copy key crashes in HeliBoard language settings text field, for any non-empty text:
```
java.lang.NegativeArraySizeException: -2
at dalvik.system.VMRuntime.newUnpaddedArray(Native Method)
at com.android.internal.util.ArrayUtils.newUnpaddedCharArray(ArrayUtils.java:51)
at android.text.SpannableStringBuilder.<init>(SpannableStringBuilder.java:63)
at androidx.emoji2.text.SpannableBuilder.<init>(SpannableBuilder.java:86)
at androidx.emoji2.text.SpannableBuilder.subSequence(SpannableBuilder.java:125)
at android.view.inputmethod.BaseInputConnection.getTextAfterCursor(BaseInputConnection.java:561)
at android.view.inputmethod.InputConnectionWrapper.getTextAfterCursor(InputConnectionWrapper.java:88)
at com.android.internal.view.IInputConnectionWrapper.executeMessage(IInputConnectionWrapper.java:246)
at com.android.internal.view.IInputConnectionWrapper.dispatchMessage(IInputConnectionWrapper.java:225)
at com.android.internal.view.IInputConnectionWrapper.getTextAfterCursor(IInputConnectionWrapper.java:117)
at com.android.internal.view.InputConnectionWrapper.getTextAfterCursor(InputConnectionWrapper.java:260)
at helium314.keyboard.latin.RichInputConnection.getTextAfterCursorAndDetectLaggyConnection(RichInputConnection.java:466)
at helium314.keyboard.latin.RichInputConnection.getTextAfterCursor(RichInputConnection.java:453)
at helium314.keyboard.latin.RichInputConnection.copyAllText(RichInputConnection.java:675)
at helium314.keyboard.latin.inputlogic.InputLogic.handleFunctionalEvent(InputLogic.java:733)
at helium314.keyboard.latin.inputlogic.InputLogic.onCodeInput(InputLogic.java:477)
at helium314.keyboard.latin.LatinIME.onEvent(LatinIME.java:1487)
at helium314.keyboard.latin.LatinIME.onCodeInput(LatinIME.java:1477)
at helium314.keyboard.latin.suggestions.SuggestionStripView.onLongClickToolKey(SuggestionStripView.java:380)
at helium314.keyboard.latin.suggestions.SuggestionStripView.onLongClick(SuggestionStripView.java:367)
at android.view.View.performLongClickInternal(View.java:6677)
at android.view.View.performLongClick(View.java:6635)
at android.view.View.performLongClick(View.java:6653)
at android.view.View$CheckForLongPress.run(View.java:25871)
at android.os.Handler.handleCallback(Handler.java:873)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:193)
at android.app.ActivityThread.main(ActivityThread.java:6718)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

> may copy up to Integer.MAX_VALUE * 2 bytes to memory

I would not care about this. If you already get that many characters into a text field, you probably need to have so much memory that you can spare those 4 GB anyway.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/clipboard/ClipboardHistoryView.kt:
```diff
@@ -240,6 +242,22 @@ class ClipboardHistoryView @JvmOverloads constructor(
+            if (longClickCode != KeyCode.UNSPECIFIED) {
+                keyboardActionListener?.onCodeInput(
+                    longClickCode,
+                    Constants.SUGGESTION_STRIP_COORDINATE,
+                    Constants.SUGGESTION_STRIP_COORDINATE,
```
Is there a specific reason you use `SUGGESTION_STRIP_COORDINATE` here? If not, I would suggest to use `NOT_A_COORDINATE` like it's done `onClick`.

Or if there is a good reason to use `SUGGESTION_STRIP_COORDINATE`, maybe it should be used in `onClick` too?

codokie replied:
but isn't it already using 'SUGGESTION_STRIP_COORDINATE' in onClick? I remember copying the line from there

Helium314 replied:
This is onClick:

```kotlin
    override fun onClick(view: View) {
        when (view) {
            alphabetKey, spacebar, deleteKey -> {
                keyboardActionListener?.onCodeInput(view.tag as Int,
                    Constants.NOT_A_COORDINATE, Constants.NOT_A_COORDINATE, false)
                keyboardActionListener?.onReleaseKey(view.tag as Int, false)
            }
        }
        val tag = view.tag
        if (tag is ToolbarKey) {
            val code = getCodeForToolbarKey(tag)
            if (code != null) {
                keyboardActionListener?.onCodeInput(code, Constants.NOT_A_COORDINATE, Constants.NOT_A_COORDINATE, false)
                return
            }
            if (tag == ToolbarKey.CLEAR_CLIPBOARD)
                clipboardHistoryManager?.clearHistory()
        }
    }
```

codokie replied:
Right sorry I copied it from the SuggestionStripView, did not realize NOT_A_COORDINATE is used for the ClipboardHistoryView. Will fix this. Thanks

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/internal/keyboard_parser/floris/KeyCode.kt:
```diff
@@ -131,6 +131,9 @@ object KeyCode {
     // Code value representing the code is not specified.
     const val NOT_SPECIFIED =             -10008 // todo: not sure if there is need to have the "old" unspecified keyCode different, just test it and maybe merge
+    const val CLIPBOARD_COPY_ALL =        -10009
+    const val MOVE_PAGE_UP =              -10010
+    const val MOVE_PAGE_DOWN =            -10011
```
I think the keys codes should be names just `PAGE_UP` / `PAGE_DOWN`. Naming is not fully consistent, but I'm trying to...

codokie replied:
i agree, it will be changed

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/clipboard/ClipboardHistoryView.kt:
```diff
@@ -240,6 +242,22 @@ class ClipboardHistoryView @JvmOverloads constructor(
+                )
+            }
+        }
+        return true
```
I think you should only return true if the key is a toolbar key, or if the toolbar key actually has a specified key code.
[Documentation](https://developer.android.com/reference/android/view/View.OnLongClickListener#onLongClick(android.view.View)) states that the return value should be "true if the callback consumed the long click, false otherwise."

Or did you already try it and it caused issues?

codokie replied:
i don't think this will have an effect because there are no views with a fallback long click function but i can change it just in case you decide to add one

Helium314 replied:
Thanks!
I prefer to stick with behaving as documented whenever (easily) possible, makes like easier when changing things later.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/suggestions/SuggestionStripView.java:
```diff
@@ -375,10 +373,13 @@ public boolean onLongClick(final View view) {
     private void onLongClickToolKey(final View view) {
-        if (ToolbarKey.CLIPBOARD == view.getTag() && view.getParent() == mPinnedKeys) {
-            onLongClickClipboardKey(); // long click pinned clipboard key
+        final ToolbarKey tag = (ToolbarKey) view.getTag();
```
Better make sure this line does not crash, e.g. you can check `instanceof` before the cast.
(I know that currently `onLongClickToolKey` is not called if the tag is not a toolbar key, but better be safe)

codokie replied:
will do

codokie commented:
> Pasting by long pressing the clipboard key replaces the underlined word, which is not the case without this PR.

right I think using onTextInput instead of commitText will fix that, but I still need to figure out how to get the suggestions to update
> Long pressing the copy key crashes in HeliBoard language settings text field, for any non-empty text:

weird, I am unable to replicate the crash..
It seems that the problem is that the RichInputConnection.getTextAfterCursor() (which I did not alter) does not catch exceptions.

Thank you for the thorough review

codokie commented:
When you've got the time please check if the copy key long click following [2fe0917](https://github.com/Helium314/HeliBoard/pull/691/commits/2fe091710a7c49b57d488b21aa2d9d4998bafa33) still crashes as I can't replicate it. Thanks

Helium314 commented on app/src/main/java/helium314/keyboard/latin/LatinIME.java:
```diff
@@ -1524,6 +1524,7 @@ public void onTextInput(final String rawText) {
                 mInputLogic.onTextInput(mSettings.getCurrent(), event,
                         mKeyboardSwitcher.getKeyboardShiftMode(), mHandler);
         updateStateAfterInputTransaction(completeInputTransaction);
+        mInputLogic.restartSuggestionsOnWordTouchedByCursor(mSettings.getCurrent(), mKeyboardSwitcher.getCurrentKeyboardScript());
```
Thanks, that's a nice improvement!

Helium314 commented:
Everything working now!
